### PR TITLE
Files and Exceptions in Python

### LINKED ISSUE #2263: Files and Exceptions in Python
### Proposed title of article

Files and Exceptions in Python

### Introduction paragraph (2-3 paragraphs):
Working with files and saving data makes your Python programs easier for people to use. Users can run your program, choose what data to enter and when to enter it, do some work, and then close the program and pick up where they left off later.

Exceptions are special objects Python creates to manage errors that arise while a program is running. Understanding exceptions will help you deal with situations in which files don’t exist and deal with other problems that can cause your programs to crash. Exceptions are handled with try - except blocks. When you use try - except blocks, your programs will continue running even if things start to go wrong. Instead of trace-backs, which can be confusing for users to read, users will see friendly error messages that you write.

In this article we will learn to work with files and how to handle errors so your Python programs don’t crash when they encounter unexpected situations.
### Key takeaways
In this article,the reader should be able to:

   1.Read from a file
   2.Work with a File’s Contents
   3.Write to a File
   4.Handle the ZeroDivisionError Exception
    5.Handle the FileNotFoundError Exception

### References
N/A

### PR DESCRIPTION
**Pull request by duncandegwa**
Files and Exceptions in Python

### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 

- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2263

### CODE REVIEW

File: content/articles/files-and-exeptions-in-python/index.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
+ 
+
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:2]: Could the introduction be more elaborative? 

Talk in-detail about Files and exceptions in Python, what their uses are, and talk about what the reader is going to learn, what are the key takeaways and outcomes.

Have a look at other Section articles.

File: content/articles/files-and-exeptions-in-python/index.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:1]: The file directory name must be `files-and-exceptions-in-python`

`exeptions` -> `exceptions`

File: content/articles/files-and-exeptions-in-python/index.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+Working with files will make your programs fast when analyzing masses of data. Exceptions are special objects that Python uses to manage errors that occur when a program is running. In this tutorial, we’ll learn working with files and handling some exception errors.
+
+### Prerequisites
+
+Have some basic knowledge of Python language.
+
+### Reading from a file
+
+Working with data in a text file, you need to read the file into memory first. To read a file, you must have a file with some data.
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:11]: `To read a file, you must have a file with some data.`

This is not true. You can read an empty file as well.

File: content/articles/files-and-exeptions-in-python/index.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+```python
+
+with open(‘years.txt’) as file_object:
+
+    contents = file_object.read()
+
+ print(contents)
+
+```
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:35]: Remove an extra new line, after each line.

Comment (srishilesh) [reply content/articles/files-and-exeptions-in-python/index.md:35]: Do this for all the code snippets

File: content/articles/files-and-exeptions-in-python/index.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+The `open()` function takes one argument which is the name of the file to be opened. Python looks for the **years.txt** file in the folder where our Python file is stored. The `open()` function returns an object representing the file (years.txt) which is then stored in  variable file_object.
+
+The keyword `with` closes the file when access to it is no longer need.
+
+The `read()` method is used to read the whole data in the file and store it in contents.
+
+When we run the above code, we get the output as:
+
+```python
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:45]:
```suggestion
```bash
```

File: content/articles/files-and-exeptions-in-python/index.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+print(len(yrs_string)) # print how long the string is 
+
+ # OUTPUT
+
+ # 209020702089
+
+ # 12
+
+```
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:83]: Remove new line

File: content/articles/files-and-exeptions-in-python/index.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+### Exceptions
+
+Exceptions are unique objects that Python uses to control errors that occur when a program is running. Exceptions errors arise when correct syntactically Python programs produce an error. When these errors arise, Python creates an exception object. When we write codes that deal with the exception, our programs will continue running. If we don’t our programs will stop executing and show a trace-back, which is very hard for a user to understand.
+
+Python uses the *try-except* block to control exceptions. A try-except block informs Python how to act when an exception emerges. Our programs will continue to execute even if things go wrong.
+
+### Handling the ZeroDivisionError exception
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:163]:
```suggestion
#### Handling the ZeroDivisionError exception
```

File: content/articles/files-and-exeptions-in-python/index.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+```python
+
+print(6/0)
+
+```
+
+When we run the above code Python gives the following trace-back:
+
+```python
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:175]:
```suggestion
```bash
```

File: content/articles/files-and-exeptions-in-python/index.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+ print(6/0)
+
+except ZeroDivisionError:
+
+ print(“You can’t divide by zero!”) # You can’t divide by zero!
+
+```
+
+### Handling the FileNotFoundError exception
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:205]:
```suggestion
#### Handling the FileNotFoundError exception
```

File: content/articles/files-and-exeptions-in-python/index.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+I love playing games
+
+```
+
+### Exceptions
+
+Exceptions are unique objects that Python uses to control errors that occur when a program is running. Exceptions errors arise when correct syntactically Python programs produce an error. When these errors arise, Python creates an exception object. When we write codes that deal with the exception, our programs will continue running. If we don’t our programs will stop executing and show a trace-back, which is very hard for a user to understand.
+
+Python uses the *try-except* block to control exceptions. A try-except block informs Python how to act when an exception emerges. Our programs will continue to execute even if things go wrong.
+
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:162]: Explain about `try-catch-except` blocks.

File: content/articles/files-and-exeptions-in-python/index.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+ - Read from a file
+
+ - Work with a File’s Contents
+
+ - Write to a File
+
+ - Handle the ZeroDivisionError Exception
+
+ - Handle the FileNotFoundError Exception
+
```
Comment (srishilesh) [inline content/articles/files-and-exeptions-in-python/index.md:265]: Add a `Further reading`/`References` section where you add external links that helps reader gain more knowledge on this topic

**Review by srishilesh — CHANGES_REQUESTED** [review]
@duncandegwa 

I've requested a few changes, please let me know when done.

Also, change the name of directory to `files-and-exceptions-in-python` (`exeptions` -> `exceptions`). And, remove extra new lines (especially in code snippets).

Thanks

**Comment by duncandegwa** [comment]
@srishilesh have done the suggested changes
Thank you

**Review by srishilesh — APPROVED** [review]
@duncandegwa 

Thanks for making the changes.

Moving for the final review. Please go through my commits to understand and improve on your future articles.

Thanks

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by louisefindlay23 — APPROVED** [review]
Congratulations, @duncandegwa. Your article has now been passed final review and been published.

cc - @srishilesh

**Pull request merged**